### PR TITLE
support all h2 jdbc urls

### LINKED ISSUE #13007: Support all the jdbc urls for H2 databases
### Is your feature request related to a problem?
Currently the [H2DataSourceMetaData](https://github.com/apache/shardingsphere/blob/master/shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java#L43) only supports the memory jdbc url like ```jdbc:h2:mem```. While I'm workong with [quarkus-shardingsphere-jdbc](https://github.com/quarkiverse/quarkus-shardingsphere-jdbc), it needs to support all the [H2 jdbc urls ](https://quarkus.io/guides/datasource#h2)

```
jdbc:h2:{ {.|mem:}[name] | [file:]fileName | {tcp|ssl}:[//]server[:port][,server2[:port]]/name }[;key=value…​]
```
### Describe the feature you would like.

It could be great to support all of these urls.

### PR DESCRIPTION
**Pull request by lichengxin813**
Fixes #13007.

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/pom.xml [hunk 1]
```diff
@@ -171,6 +171,7 @@
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>
                 <artifactId>templating-maven-plugin</artifactId>
+                <version>1.0.0</version>
```
Comment (sandynz) [inline shardingsphere-proxy/shardingsphere-proxy-backend/pom.xml:174]: It's not related with this issue, should not be committed.

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java [hunk 2]
```diff
@@ -42,9 +42,31 @@ public void assertNewConstructorWithSymbol() {
         assertThat(actual.getPort(), is(-1));
         assertNull(actual.getSchema());
     }
-    
-    @Test(expected = UnrecognizedDatabaseURLException.class)
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java:46]: `UnrecognizedDatabaseURLException` might be not used anymore, importing could be removed.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 3]
```diff
@@ -39,17 +39,21 @@
...
     public H2DataSourceMetaData(final String url) {
         Matcher matcher = pattern.matcher(url);
         if (!matcher.find()) {
             throw new UnrecognizedDatabaseURLException(url, pattern.pattern());
         }
+        String portFromMatcher = matcher.group("port");
+        String catalogFromMatcher = matcher.group("catalog");
+        String nameFromMatcher = matcher.group("name");
+        String fileNameFromMatcher = matcher.group("fileName");
         hostName = "";
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:54]: `hostName` could be set a value if it's `tcp/ssl` mode.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 4]
```diff
@@ -39,17 +39,21 @@
...
-        port = DEFAULT_PORT;
-        catalog = matcher.group(2);
+        port =  portFromMatcher == null || portFromMatcher.isEmpty() ? DEFAULT_PORT : Integer.parseInt(portFromMatcher);
+        catalog = catalogFromMatcher == null ? (nameFromMatcher == null ? fileNameFromMatcher : nameFromMatcher) : catalogFromMatcher;
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:56]: According to the code of conduct, `null` (constant) should be at left of variable.

File: shardingsphere-proxy/shardingsphere-proxy-backend/pom.xml (hunk 1 above)
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/pom.xml:174]: Could you revert this irrelevant change?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 5]
```diff
@@ -39,17 +39,21 @@
     private final String catalog;
     
     private final String schema;
-    
-    private final Pattern pattern = Pattern.compile("jdbc:h2:(mem|~)[:/]([\\w\\-]+);?\\S*", Pattern.CASE_INSENSITIVE);
+
+    private final Pattern pattern = Pattern.compile("jdbc:h2:((mem|~)[:/](?<catalog>[\\w\\-]+)|(ssl:|tcp:)(//)?([\\w\\-.]+)(:(?<port>[0-9]*)/)?([/~\\w\\-.]+)(/(?<name>[\\-\\w]*)){1}|file:([/~\\w\\-]+)(/(?<fileName>[\\-\\w]*)){1});?\\S*", Pattern.CASE_INSENSITIVE);
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:43]: Could `*` in `<port>[0-9]*` be `{1,4}`?
And also other `*`.

**Review by sandynz — CHANGES_REQUESTED** [review]

**Review by sandynz — APPROVED** [review]

**Pull request merged**